clients/drcachesim/tests/raw2trace_unit_tests.cpp

```diff
@@ -1508,6 +1508,82 @@ test_rseq_rollback_with_signal(void *drcontext)
         check_entry(entries, idx, TRACE_TYPE_FOOTER, -1));
 }
 
+/* Tests rseq rollback with taken branch right before signal. */
+bool
+test_rseq_rollback_with_signal_after_branch(void *drcontext)
+{
+    std::cerr << "\n===============\nTesting legacy rseq rollback with taken branch "
+                 "before signal.\n";
+    instrlist_t *ilist = instrlist_create(drcontext);
+    // raw2trace doesn't like offsets of 0 so we shift with a nop.
+    instr_t *nop = XINST_CREATE_nop(drcontext);
+    instr_t *move1 =
+        XINST_CREATE_move(drcontext, opnd_create_reg(REG1), opnd_create_reg(REG2));
+    instr_t *store =
+        XINST_CREATE_store(drcontext, OPND_CREATE_MEMPTR(REG2, 0), opnd_create_reg(REG1));
+    instr_t *jcc =
+        XINST_CREATE_jump_cond(drcontext, DR_PRED_EQ, opnd_create_instr(store));
+    instr_t *move2 =
+        XINST_CREATE_move(drcontext, opnd_create_reg(REG2), opnd_create_reg(REG1));
+    instrlist_append(ilist, nop);
+    instrlist_append(ilist, jcc);
+    instrlist_append(ilist, move1);
+    instrlist_append(ilist, store);
+    instrlist_append(ilist, move2);
+    size_t offs_nop = 0;
+    size_t offs_jcc = offs_nop + instr_length(drcontext, nop);
+    size_t offs_move1 = offs_jcc + instr_length(drcontext, jcc);
+    size_t offs_store = offs_move1 + instr_length(drcontext, move1);
+    size_t offs_move2 = offs_store + instr_length(drcontext, store);
+
+    std::vector<offline_entry_t> raw;
+    raw.push_back(make_header());
+    raw.push_back(make_tid());
+    raw.push_back(make_pid());
+    raw.push_back(make_line_size());
+    raw.push_back(make_timestamp());
+    raw.push_back(make_core());
+    raw.push_back(make_marker(TRACE_MARKER_TYPE_RSEQ_ENTRY, offs_move2));
+    raw.push_back(make_block(offs_jcc, 1));
+    raw.push_back(make_marker(TRACE_MARKER_TYPE_RSEQ_ABORT, offs_store));
+    raw.push_back(make_marker(TRACE_MARKER_TYPE_KERNEL_EVENT, offs_store));
+    raw.push_back(make_block(offs_move2, 1));
```

Comment:
It is confusing to have the rseq abort due to signal followed by the post-rseq instr: normally it would go to a handler with another KERNEL_EVENT?  If the user marked it SIG_IGN: it wouldn't have interrupted at all.  Wouldn't it always go to a handler then?

I see some other tests doing this: at least add a comment saying this is unusual but simplfiied for test purposes?

Reply:
Yeah, it's a simplified version of traces from code_api|tool.drcacheoff.rseq:

      141487      105678:     3662080 ifetch       7 byte(s) @ 0x00007fe7abdfa938 80 3d b9 38 00 00 00 cmp    <rel> 0x00007fe7abdfe1f8, $0x00
      141488      105678:     3662080 read         1 byte(s) @ 0x00007fe7abdfe1f8 by PC 0x00007fe7abdfa938
      141489      105679:     3662080 ifetch       2 byte(s) @ 0x00007fe7abdfa93f 74 0a                jz     $0x00007fe7abdfa94b (untaken)
      141490      105680:     3662080 ifetch       8 byte(s) @ 0x00007fe7abdfa941 0f 18 1c 25 01 00 00 prefetcht2 0x01
      141490      105680:     3662080                                             00
      141491      105680:     3662080 pref-r-L3    1 byte(s) @ 0x0000000000000001 by PC 0x00007fe7abdfa941
      141492      105681:     3662080 ifetch       2 byte(s) @ 0x00007fe7abdfa949 0f 0b                ud2
      141493      105681:     3662080 <marker: rseq abort from 0x7fe7abdfa949 to handler>
      141494      105681:     3662080 <marker: kernel xfer from 0x7fe7abdfa949 to handler>
      141495      105681:     3662080 <marker: timestamp 13351125423049607>
      141496      105681:     3662080 <marker: tid 3662080 on core 4110>
      141497      105681:     3662080 <marker: kernel xfer from 0x7fe7abdfa957 to handler>
      141498      105681:     3662080 <marker: timestamp 13351125423049609>
      141499      105681:     3662080 <marker: tid 3662080 on core 4110>

I've added a comment accordingly.